common/cache.go

```diff
@@ -10,87 +13,169 @@ import (
+	// Invariant: m[FirstBlock..NextBlock) are valid.
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
 
-	if height > c.nextBlock {
-		// restarting the cache (never happens currently), or first time
-		for i := c.firstBlock; i < c.nextBlock; i++ {
-			delete(c.m, i)
-		}
-		c.firstBlock = height
-		c.nextBlock = height
+	height := int(block.Height)
+	if height != c.NextBlock {
+		Log.Fatalf("bug, cache.Add non-consecutive blocks: height: %v, ccache: %+v", height, c)
 	}
-	// Invariant: m[firstBlock..nextBlock) are valid.
 
-	// If we already have this block, a reorg must have occurred;
-	// this block (and all higher) must be re-added.
-	h := height
-	if h < c.firstBlock {
-		h = c.firstBlock
+	// Add the new block and its length to the db files
+	data, err := proto.Marshal(block)
+	if err != nil {
+		return err
 	}
-	for i := h; i < c.nextBlock; i++ {
-		delete(c.m, i)
+	_, err = c.blocksFile.Write(data)
+	if err != nil {
+		Log.Fatal("blocks write failed: ", err)
 	}
-	c.nextBlock = height
-	if c.firstBlock > c.nextBlock {
-		c.firstBlock = c.nextBlock
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, uint32(len(data)))
+	_, err = c.lengthsFile.Write(b)
+	if err != nil {
+		Log.Fatal("lengths write failed: ", err)
 	}
-	// Invariant: m[firstBlock..nextBlock) are valid.
 
-	// Detect reorg, ingestor needs to handle it
-	if height > c.firstBlock && !bytes.Equal(block.PrevHash, c.m[height-1].hash) {
-		return true, nil
+	// update the in-memory variables
+	offset := c.starts[len(c.starts)-1]
+	c.starts = append(c.starts, offset+int64(len(data)))
+
+	if c.LatestHash == nil {
+		c.LatestHash = make([]byte, len(block.Hash))
 	}
+	copy(c.LatestHash, block.Hash)
+	c.NextBlock++
+	// Invariant: m[FirstBlock..NextBlock) are valid.
+	return nil
+}
 
-	// Add the entry and update the counters
-	data, err := proto.Marshal(block)
-	if err != nil {
-		return false, err
+// Reorg resets NextBlock (the block that should be Add()ed next) to the given
+// height, taking care of all the details. It returns the height that should be
+// added next (different if less than FirstBlock).
+func (c *BlockCache) Reorg(height int) int {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if height < c.FirstBlock {
+		height = c.FirstBlock
 	}
-	c.m[height] = &blockCacheEntry{
-		data: data,
-		hash: block.GetHash(),
+	if height > c.NextBlock {
+		Log.Fatalf("bug, cache.Reorg moves ahead: height: %v, ccache: %+v", height, c)
 	}
-	c.nextBlock++
-	// Invariant: m[firstBlock..nextBlock) are valid.
+	// Remove the end of the cache.
+	c.NextBlock = height
+	newCacheLen := height - c.FirstBlock
+	c.starts = c.starts[:newCacheLen+1]
 
-	// remove any blocks that are older than the capacity of the cache
-	for c.firstBlock < c.nextBlock-c.MaxEntries {
-		// Invariant: m[firstBlock..nextBlock) are valid.
-		delete(c.m, c.firstBlock)
-		c.firstBlock++
+	if err := c.lengthsFile.Truncate(int64(4 * newCacheLen)); err != nil {
+		Log.Fatal("truncate failed: ", err)
```

Comment:
This could also leave the database files in a state where there's more blocks in the blocks file than there are lengths in the lengths file and lead to similar bugs as above. The same fix works here too.